In [1]:
import pandas as pd

In [2]:
from sklearn import linear_model

In [3]:
climateInfo = pd.read_csv('climate_change.csv')

In [4]:
climateInfo.describe()

Year       Month         MEI         CO2          CH4  \
count   308.000000  308.000000  308.000000  308.000000   308.000000   
mean   1995.662338    6.551948    0.275555  363.226753  1749.824513   
std       7.423197    3.447214    0.937918   12.647125    46.051678   
min    1983.000000    1.000000   -1.635000  340.170000  1629.890000   
25%    1989.000000    4.000000   -0.398750  353.020000  1722.182500   
50%    1996.000000    7.000000    0.237500  361.735000  1764.040000   
75%    2002.000000   10.000000    0.830500  373.455000  1786.885000   
max    2008.000000   12.000000    3.001000  388.500000  1814.180000   

              N2O      CFC-11      CFC-12          TSI    Aerosols        Temp  
count  308.000000  308.000000  308.000000   308.000000  308.000000  308.000000  
mean   312.391834  251.973068  497.524782  1366.070759    0.016657    0.256776  
std      5.225131   20.231783   57.826899     0.399610    0.029050    0.179090  
min    303.677000  191.324000  350.113000  1365.426100    0.001600   -0.282000  
25%    308.111500  246.295500  472.410750  1365.717050    0.002800    0.121750  
50%    311.507000  258.344000  528.356000  1365.980900    0.005750    0.248000  
75%    316.979000  267.031000  540.524250  1366.363250    0.012600    0.407250  
max    322.182000  271.494000  543.813000  1367.316200    0.149400    0.739000

In [5]:
climateInfo.head()

Year  Month    MEI     CO2      CH4      N2O   CFC-11   CFC-12        TSI  \
0  1983      5  2.556  345.96  1638.59  303.677  191.324  350.113  1366.1024   
1  1983      6  2.167  345.52  1633.71  303.746  192.057  351.848  1366.1208   
2  1983      7  1.741  344.15  1633.22  303.795  192.818  353.725  1366.2850   
3  1983      8  1.130  342.25  1631.35  303.839  193.602  355.633  1366.4202   
4  1983      9  0.428  340.17  1648.40  303.901  194.392  357.465  1366.2335   

   Aerosols   Temp  
0    0.0863  0.109  
1    0.0794  0.118  
2    0.0731  0.137  
3    0.0673  0.176  
4    0.0619  0.149

In [6]:
climateInfo.isnull().any()

Year        False
Month       False
MEI         False
CO2         False
CH4         False
N2O         False
CFC-11      False
CFC-12      False
TSI         False
Aerosols    False
Temp        False
dtype: bool

#### Split the data into a training set and a test set. Training set will have all the observations up to and including 2006. Test set will have all the observations after 2006.

In [8]:
climateInfoUntil2006 = climateInfo[climateInfo['Year'] <= 2006]

In [9]:
climateInfoUntil2006.describe()

Year       Month         MEI         CO2          CH4  \
count   284.000000  284.000000  284.000000  284.000000   284.000000   
mean   1994.661972    6.556338    0.341923  361.414261  1745.841479   
std       6.845996    3.446768    0.929639   11.439691    45.669846   
min    1983.000000    1.000000   -1.586000  340.170000  1629.890000   
25%    1989.000000    4.000000   -0.323000  352.315000  1716.347500   
50%    1995.000000    7.000000    0.308500  359.890000  1758.605000   
75%    2001.000000   10.000000    0.898000  370.585000  1781.637500   
max    2006.000000   12.000000    3.001000  384.980000  1808.150000   

              N2O      CFC-11      CFC-12          TSI    Aerosols        Temp  
count  284.000000  284.000000  284.000000   284.000000  284.000000  284.000000  
mean   311.657225  252.487092  494.217546  1366.101437    0.017721    0.247799  
std      4.758513   20.987671   59.046642     0.401283    0.030014    0.181136  
min    303.677000  191.324000  350.113000  1365.426100    0.001600   -0.282000  
25%    307.657000  249.557750  462.543000  1365.754550    0.002700    0.118000  
50%    310.849500  260.373500  522.089000  1366.054500    0.006200    0.232500  
75%    316.129250  267.448000  540.972750  1366.399275    0.014000    0.406500  
max    320.451000  271.494000  543.813000  1367.316200    0.149400    0.739000

#### build a linear regression model to predict the dependent variable Temp, using MEI, CO2, CH4, N2O, CFC.11, CFC.12, TSI, and Aerosols as independent variables.

In [10]:
X_train_until_2006 = climateInfoUntil2006[['MEI', 'CO2', 'CH4', 'N2O', 'CFC-11', 'CFC-12', 'TSI', 'Aerosols']]
y_train_until_2006 = climateInfoUntil2006[['Temp']]

In [11]:
linear_regression = linear_model.LinearRegression()

In [12]:
linear_regression.fit(X_train_until_2006, y_train_until_2006)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=1, normalize=False)

In [13]:
linear_regression.coef_

array([[  6.42053134e-02,   6.45735927e-03,   1.24041896e-04,
         -1.65280033e-02,  -6.63048889e-03,   3.80810324e-03,
          9.31410835e-02,  -1.53761324e+00]])

In [14]:
linear_regression.intercept_

array([-124.5942604])

In [15]:
linear_regression.score(X_train_until_2006, y_train_until_2006)

0.75089327705234177

\begin{equation*}
R^2   = 0.75
\end{equation*}

***

In [18]:
from sklearn import feature_selection

In [19]:
feature_selection.f_regression(X_train_until_2006, y_train_until_2006)

C:\Users\Vivek\Anaconda3\lib\site-packages\sklearn\utils\validation.py:526: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


(array([   8.64559024,  463.59447651,  275.9381421 ,  434.2424694 ,
          56.22183027,  252.83581202,   17.75609183,   49.04753697]),
 array([  3.55017288e-03,   1.74154020e-61,   1.10728523e-43,
          5.07880389e-59,   8.44135858e-13,   4.40016348e-41,
          3.38363685e-05,   1.83341730e-11]))

#### Which variables are significant in the model? We will consider a variable signficant only if the p-value is below 0.05. 

- MEI
- C02
- CFC-11
- CFC-12
- TSI
- Aerosols

In [20]:
import statsmodels.api as sm

X = sm.add_constant(X_train_until_2006)

model = sm.OLS(y_train_until_2006, X).fit()

model.summary()

C:\Users\Vivek\Anaconda3\lib\site-packages\statsmodels\compat\pandas.py:56: FutureWarning: The pandas.core.datetools module is deprecated and will be removed in a future version. Please use the pandas.tseries module instead.
  from pandas.core import datetools


<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   Temp   R-squared:                       0.751
Model:                            OLS   Adj. R-squared:                  0.744
Method:                 Least Squares   F-statistic:                     103.6
Date:                Sun, 14 Oct 2018   Prob (F-statistic):           1.94e-78
Time:                        07:32:46   Log-Likelihood:                 280.10
No. Observations:                 284   AIC:                            -542.2
Df Residuals:                     275   BIC:                            -509.4
Df Model:                           8                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const       -124.5943     19.887     -6.265      0.000    -163.744     -85.445
MEI            0.0642      0.006      9.923      0.000       0.051       0.077
CO2            0.0065      0.002      2.826      0.005       0.002       0.011
CH4            0.0001      0.001      0.240      0.810      -0.001       0.001
N2O           -0.0165      0.009     -1.930      0.055      -0.033       0.000
CFC-11        -0.0066      0.002     -4.078      0.000      -0.010      -0.003
CFC-12         0.0038      0.001      3.757      0.000       0.002       0.006
TSI            0.0931      0.015      6.313      0.000       0.064       0.122
Aerosols      -1.5376      0.213     -7.210      0.000      -1.957      -1.118
==============================================================================
Omnibus:                        8.740   Durbin-Watson:                   0.956
Prob(Omnibus):                  0.013   Jarque-Bera (JB):               10.327
Skew:                           0.289   Prob(JB):                      0.00572
Kurtosis:                       3.733   Cond. No.                     8.53e+06
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 8.53e+06. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

***

#### Current scientific opinion is that nitrous oxide and CFC-11 are greenhouse gases: gases that are able to trap heat from the sun and contribute to the heating of the Earth. However, the regression coefficients of both the N2O and CFC-11 variables are negative, indicating that increasing atmospheric concentrations of either of these two compounds is associated with lower global temperatures.

#### Which of the following is the simplest correct explanation for this contradiction?
1. Climate scientists are wrong that N2O and CFC-11 are greenhouse gases - this regression analysis constitutes part of a disproof.
2. There is not enough data, so the regression coefficients being estimated are not accurate.
3. All of the gas concentration variables reflect human development - N2O and CFC.11 are correlated with other variables in the data set.

#### Answer: 3

#### Explanation:
#### The linear correlation of N2O and CFC.11 with other variables in the data set is quite large. The first explanation does not seem correct, as the warming effect of nitrous oxide and CFC-11 are well documented, and our regression analysis is not enough to disprove it. The second explanation is unlikely, as we have estimated eight coefficients and the intercept from 284 observations.

***

In [21]:
X_train_until_2006.corr()

MEI       CO2       CH4       N2O    CFC-11    CFC-12  \
MEI       1.000000 -0.041147 -0.033419 -0.050820  0.069000  0.008286   
CO2      -0.041147  1.000000  0.877280  0.976720  0.514060  0.852690   
CH4      -0.033419  0.877280  1.000000  0.899839  0.779904  0.963616   
N2O      -0.050820  0.976720  0.899839  1.000000  0.522477  0.867931   
CFC-11    0.069000  0.514060  0.779904  0.522477  1.000000  0.868985   
CFC-12    0.008286  0.852690  0.963616  0.867931  0.868985  1.000000   
TSI      -0.154492  0.177429  0.245528  0.199757  0.272046  0.255303   
Aerosols  0.340238 -0.356155 -0.267809 -0.337055 -0.043921 -0.225131   

               TSI  Aerosols  
MEI      -0.154492  0.340238  
CO2       0.177429 -0.356155  
CH4       0.245528 -0.267809  
N2O       0.199757 -0.337055  
CFC-11    0.272046 -0.043921  
CFC-12    0.255303 -0.225131  
TSI       1.000000  0.052117  
Aerosols  0.052117  1.000000

#### Compute the correlations between all the variables in the training set. Which of the independent variables is N2O highly correlated with (absolute correlation greater than 0.7).
- C02
- CH4
- CFC-12


#### Which of the independent variables is CFC.11 highly correlated with?
- CH4
- CFC-12

#### Given that the correlations are so high, let us focus on the N2O variable and build a model with only MEI, TSI, Aerosols and N2O as independent variables. Remember to use the training set to build the model.

In [22]:
X_train_N2O = climateInfoUntil2006[['MEI', 'TSI', 'Aerosols', 'N2O']]

In [23]:
linear_regression.fit(X_train_N2O, y_train_until_2006)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=1, normalize=False)

In [24]:
linear_regression.coef_

array([[ 0.06418576,  0.07949028, -1.70173707,  0.02531975]])

In [25]:
linear_regression.score(X_train_N2O, y_train_until_2006)

0.72613212795111437

#### What is the coefficient of N2O in this reduced model?

0.02531975

(How does this compare to the coefficient in the previous model with all of the variables?)

Enter the model ** R^2 **: *0.72613212795111437*

#### Explanation:
#### We have observed that, for this problem, when we remove many variables the sign of N2O flips. The model has not lost a lot of explanatory power (the model R^2 is 0.7261 compared to 0.7509 previously) despite removing many variables. This type of behavior is typical when building a model where many of the independent variables are highly correlated with each other. In this particular problem many of the variables (CO2, CH4, N2O, CFC.11 and CFC.12) are highly correlated, since they are all driven by human industrial development.

***

#### We have many variables in this problem, and as we have seen above, dropping some from the model does not decrease model quality. R provides a function, step, that will automate the procedure of trying different combinations of variables to find a good compromise of model simplicity and R2. This trade-off is formalized by the Akaike information criterion (AIC) - it can be informally thought of as the quality of the model with a penalty for the number of variables in the model.

<font color = 'Blue'>Stepwise regression or Forward selection should be used for this </font>

***

#### We have developed an understanding of how well we can fit a linear regression to the training data, but does the model quality hold when applied to unseen data?

#### Using the model produced from the step function, calculate temperature predictions for the testing data set, using the predict function.

#### What is the testing set R2?

In [26]:
X_train_after_stepwise = climateInfo[['MEI', 'CO2', 'N2O', 'CFC-11', 'CFC-12', 'TSI', 'Aerosols']]
y_train = climateInfo[['Temp']]

In [40]:
linear_regression.fit(X_train_after_stepwise, y_train)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=1, normalize=False)

In [28]:
linear_regression.coef_

array([[ 0.06624108,  0.00517642, -0.01662986, -0.00726059,  0.00429837,
         0.09581909, -1.58299668]])

In [29]:
linear_regression.intercept_

array([-127.6250267])

In [39]:
linear_regression.score(X_train_after_stepwise, y_train)

0.74397992690010106

In [30]:
climateInfoAfter2006 = climateInfo[climateInfo['Year'] > 2006]

In [31]:
climateInfoAfter2006.head()

Year  Month    MEI     CO2      CH4      N2O   CFC-11   CFC-12  \
284  2007      1  0.974  382.93  1799.66  320.561  248.372  539.206   
285  2007      2  0.510  383.81  1803.08  320.571  248.264  538.973   
286  2007      3  0.074  384.56  1803.10  320.548  247.997  538.811   
287  2007      4 -0.049  386.40  1802.11  320.518  247.574  538.586   
288  2007      5  0.183  386.58  1795.65  320.445  247.224  538.130   

           TSI  Aerosols   Temp  
284  1365.7173    0.0054  0.601  
285  1365.7145    0.0051  0.498  
286  1365.7544    0.0045  0.435  
287  1365.7228    0.0045  0.466  
288  1365.6932    0.0041  0.372

In [32]:
climateInfoAfter2006.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 24 entries, 284 to 307
Data columns (total 11 columns):
Year        24 non-null int64
Month       24 non-null int64
MEI         24 non-null float64
CO2         24 non-null float64
CH4         24 non-null float64
N2O         24 non-null float64
CFC-11      24 non-null float64
CFC-12      24 non-null float64
TSI         24 non-null float64
Aerosols    24 non-null float64
Temp        24 non-null float64
dtypes: float64(9), int64(2)
memory usage: 2.2 KB


In [33]:
X_test = climateInfoAfter2006[['MEI', 'CO2', 'N2O', 'CFC-11', 'CFC-12', 'TSI', 'Aerosols']]

In [36]:
y_test = climateInfoAfter2006[['Temp']]

In [43]:
y_pred = linear_regression.predict(X_test)

In [35]:
from sklearn.metrics import mean_squared_error, r2_score

In [44]:
#Mean squared error
print("Mean squared error ", mean_squared_error(y_test, y_pred))

Mean squared error  0.00842994374592


In [47]:
#Explained variance score: 1 is perfect prediction
print('Variance score - r2 score', r2_score(y_test, y_pred))

Variance score - r2 score 0.243696871511


#### It looks like there is some issue. The model has not performed well on the test set. It should be analyzed further. 

#### It should have been 0.6286051.

***

#### Resources:

- http://scikit-learn.org/stable/auto_examples/linear_model/plot_ols.html
- https://becominghuman.ai/stats-models-vs-sklearn-for-linear-regression-f19df95ad99b
- https://datascienceplus.com/linear-regression-in-python-predict-the-bay-areas-home-prices/
- https://stackoverflow.com/questions/15433372/stepwise-regression-in-python
- https://towardsdatascience.com/train-test-split-and-cross-validation-in-python-80b61beca4b6
- https://chrisalbon.com/machine_learning/linear_regression/linear_regression_scikitlearn/
- https://www.datascience.com/learn-data-science/fundamentals/introduction-to-correlation-python-data-science
- https://stackoverflow.com/questions/27928275/find-p-value-significance-in-scikit-learn-linearregression
- http://scikit-learn.org/stable/modules/generated/sklearn.feature_selection.f_regression.html
